contracts/ControllerSolidlyLPStrat.sol

```diff
@@ -0,0 +1,271 @@
+        require(_routes[0].from == _reward, "First route isn't reward");
+        require(_routes[_routes.length - 1].to == address(want), "Last route isn't want token");
+
+        bool newReward = true;
+
+        for (uint i = 0; i < rewardTokens.length; i++) {
+            if (rewardTokens[i] == _reward) {
+                newReward = false;
+                break;
+            }
+        }
+
+        if (newReward) { rewardTokens.push(_reward); }
+
+        delete rewardToWantSolidlyRoute[_reward];
+
+        for (uint i = 0; i < _routes.length; i++) {
+            rewardToWantSolidlyRoute[_reward].push(_routes[i]);
+        }
+    }
+
+    function _deposit() internal override {
+        uint _balance = wantBalance();
+
+        if (_balance > 0) {
+            want.safeApprove(address(swapper), _balance);
+
+            (uint _amount0, uint _amount1) = swapper.swapWantForLpTokens(_balance);
+            // just in case
+            _removeAllowance(address(want), address(swapper));
+
+            _addLiquidity(_amount0, _amount1);
+        }
+    }
+
+    // amount is the want expected to be withdrawn
+    function _withdraw(uint _amount) internal override returns (uint) {
+        uint _balance = wantBalance();
+
+        if (_balance < _amount) {
+            uint _liquidity = swapper.wantToLP(_amount);
+
+            _withdrawFromPool(_liquidity);
+            _swapLPTokensForWant();
+        }
+
+        uint _withdrawn = wantBalance() - _balance;
+
+        return (_withdrawn > _amount) ? _amount : _withdrawn;
+    }
+
+    function _withdrawAll() internal override returns (uint) {
+        uint _balance = wantBalance();
+        uint _liquidity = balanceOfPool();
+
+        if (_liquidity > 0) {
+            _withdrawFromPool(_liquidity);
+            _swapLPTokensForWant();
+        }
+
+        return wantBalance() - _balance;
+    }
+
+    function _claimRewards() internal override {
+        ISolidlyGauge(gauge).getReward(address(this), rewardTokens);
+    }
+
+    function _swapRewards() internal override virtual {
+        for (uint i = 0; i < rewardTokens.length; i++) {
+            address _rewardToken = rewardTokens[i];
+            uint _balance = IERC20Metadata(_rewardToken).balanceOf(address(this));
+
+            if (_balance > 0) {
+                uint _expected = _expectedForSwap(_balance, _rewardToken, address(want));
+
+                // Want price sometimes is too high so it requires a lot of rewards to swap
+                if (_expected > 1) {
+                    IERC20Metadata(_rewardToken).safeApprove(exchange, _balance);
+
+                    ISolidlyRouter(exchange).swapExactTokensForTokens(
+                        _balance, _expected, rewardToWantSolidlyRoute[_rewardToken], address(this), block.timestamp + 60
+                    );
+                }
+            }
+        }
+    }
+
+    function _addLiquidity(uint _amount0, uint _amount1) internal {
```

Comment:
I think it'ld be better to get `amount[0|1]` from the the balance directly instead. Same in removeLiquidity.